text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1020 @@
+here will be the fact the modifiers run whenever a node is inserted or removed,
+so they may run multiple times during a components life (if they exist in an
+`{{if}}` or `{{each}}` block, for instance), and that they can receive pure
+function helpers if they do not require any state beyond the state of the
+element itself, which can simplify component code.
+
+### Converting From Classic Components
+
+For users who are familiar with the classic component API, some changes will be
+fairly simple, and others will be more nuanced. The individual focus points
+should be those outlined in the motivation of this RFC, and the motivation for
+the changes should be included in a summary that helps users understand the
+intent of new APIs so they can use them effectively.
+
+**Outer HTML semantics** should be fairly simple to communicate. These should
+also be the easiest to codemod, since it is mostly a 1-to-1 conversion from the
+classic component definition to the template.
+
+**Namespaced arguments** will be a sticking point for some. The main thing to
+focus on here is aliasing to provide defaults, and the benefits of immutability.
+Patterns like the `@arg` decorator in
+[sparkles-decorators](https://github.com/gossi/sparkles-decorators) make this
+fairly ergonomic, and for simpler components defaulting using a class field
+initializer can also work.
+
+**Lifecycle hooks** may be fairly confusing to existing users, who are used to
+having a variety of hooks to choose from when coordinating lifecycle events. The
+bullet points here are:
+
+* Tracked properties/computed properties are the primary place to react to
+  argument changes for any values that can be computed directly via getters.
+  Ideally, most logic for derived state is conventionally in tracked or computed
+  properties.
+* `didUpdateArgs` is the place to react to argument changes that don't fit into
+  getters. This includes asynchronous calls, conditional changes (e.g. if an arg
+  changed, do X, else do nothing), etc.
+* `willDestroy` is the correct place for all teardown code, like in classic
+  components.
+* `didReceiveAttrs`, `willRender`, and `didRender` code should be extracted into
+  functions which are then passed to `{{did-render}}`
+* `willDestroyElement` code should be extracted into functions which are then
+  passed to `{{will-destroy}}`
+
+**Element modifiers** will be new to almost everyone, and this will have to be
+a primary focus in learning. The bullet points for these are:
+
+* They are conceptually like the render lifecycle hooks on classic components,
+  but they run on elements instead of components.
+* `{{did-render}}` is similar to the `didRender` lifecycle hooks in classic
+  components. It will run when the element is inserted, and then it will run
+  again when any of its _args_ change.
+* `{{did-render}}` will _not_ render when sub-elements or contents of the
+  element change. If users need to react to changes in content, they should pass
```

Comment:
This again throws a wrench in the "zero DOM components" story. These components may _absolutely_ want some sort of "my children just changed" hook that doesn't require new attrs to be passed in from the outside (i.e., maybe inner components are changing their own state and re-rendering due to the google maps API returning some data)

Reply:
In looking at `ember-leaflet` and the way it works internally, I don't believe that is the case. It seems to manage most of its own state, and actually implements its own hooks. I've addressed this locally and added an example which reproduces `ember-leaflet` in a minimal fashion, will be pushing shortly.